### PR TITLE
LPS-145480 Dynamic collection is not working with combined segments

### PR DESCRIPTION
**Pull request by YurenaC**
**Motivation**
=======

Impedibug fixing: Dynamic collections are not properly working with combined segments in content display

**Proposed Solution**
=======

Review how the content display recovers all the items to be show related to the user and segments where they belongs

**Steps to Verify:**
=======

1. Go to Site Builder > Collections
2. Create a Dynamic Collection
3. Create some variations with different content and for the previous segments
4. Create a content page with Asset Publisher and another with Collection Display widgets
5. Go to System Settings > Assets > Collections
6. Enable the combine segments option for Dynamic Collections
7. Go to the content pages
8. Only the variation's content where the user belongs is shown. Also, the content only appear once.

**Change summary:**
=======

LPS-145480 Fix in dynamic collection to show combined assets that belongs only to user belonging segments order by priority

### CODE REVIEW

**Comment by YurenaC** [comment]
ci:test:sf

**Comment by YurenaC** [comment]
ci:test:relevant

File: modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java [hunk 1]
```diff
@@ -695,16 +697,44 @@ private int _dynamicSearchCount(
...
+
+		Stream<AssetListEntrySegmentsEntryRel>
+			assetListEntrySegmentsEntryRelStream = longStream.mapToObj(
+				segmentsEntryId ->
+					_assetListEntrySegmentsEntryRelLocalService.
+						fetchAssetListEntrySegmentsEntryRel(
+							assetListEntry.getAssetListEntryId(),
+							segmentsEntryId));
+
+		List<Long> filteredSegmentEntryIds =
```
Comment (cgoncas) [inline modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:721]: Maybe you could concat from longStream.mapToObj( to ).toArray();?

Comment (cgoncas) [reply modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:721]: Maybe we could avoid the collect and stream again?

Comment (YurenaC) [reply modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:721]: Absolutelly

File: modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java [hunk 2]
```diff
@@ -695,16 +696,36 @@ private int _dynamicSearchCount(
...
-	private long[] _getCombinedSegmentsEntryIds(long[] segmentEntryIds) {
+	private long[] _getCombinedSegmentsEntryIds(
+		AssetListEntry assetListEntry, long[] segmentEntryIds) {
+
 		if ((segmentEntryIds.length > 1) &&
 			ArrayUtil.contains(
 				segmentEntryIds, SegmentsEntryConstants.ID_DEFAULT)) {
 
-			return ArrayUtil.remove(
+			segmentEntryIds = ArrayUtil.remove(
```
Comment (cgoncas) [inline modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:706]: could we make this list inmutable someway?

Comment (YurenaC) [reply modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:706]: The array returned by ArrayUtil.remove(...) is already immutable as it's created specifying the length of it.

Internally, that remove treats the given array as a List and later on transform it again to an array with the following method:

`public static long[] toArray(Long[] array) {
		long[] newArray = new long[array.length];

		for (int i = 0; i < array.length; i++) {
			newArray[i] = array[i].longValue();
		}

		return newArray;
	}`

Comment (cgoncas) [reply modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:706]: Hey, but we have muted segmentEntryIds, that is a parameter, right?, or I am not getting something right?

File: modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java [hunk 3]
```diff
@@ -695,16 +696,36 @@ private int _dynamicSearchCount(
...
+
+		Stream<AssetListEntrySegmentsEntryRel>
+			assetListEntrySegmentsEntryRelStream = longStream.mapToObj(
+				segmentsEntryId ->
+					_assetListEntrySegmentsEntryRelLocalService.
+						fetchAssetListEntrySegmentsEntryRel(
+							assetListEntry.getAssetListEntryId(),
+							segmentsEntryId));
+
+		return assetListEntrySegmentsEntryRelStream.filter(
```
Comment (cgoncas) [inline modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:720]: maybe we can avoid declare assetListEntrySegmentsEntryRelStream, or SF force to declare it?

Comment (YurenaC) [reply modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java:720]: Apparently SF does not comply about it, just reworded to make it simple

**Comment by YurenaC** [comment]
@cgoncas I am dong all those updates/changes in https://github.com/liferay-tango/liferay-portal/pull/2139 so we can send the change and the test together.

I proceed to close this PR

**Pull request closed without merging**